### PR TITLE
Enable caching for backend calls

### PR DESCRIPTION
**Pull request by jquick**
This PR enables caching for all calls to the train backend. Currently this is set on the profile level. This was inspired by @sdelano and his work via https://github.com/chef/inspec/pull/2158

Signed-off-by: Jared Quick <jquick@chef.io>

### CODE REVIEW

File: lib/inspec/profile.rb
Comment (adamleff) [inline lib/inspec/profile.rb]: `||` logic with default boolean values always makes me nervous. If someone decides to make this default behavior and just changes the `false` to `true`, it will never be able to be set to false:

```
[1] pry(main)> a = true
=> true
[2] pry(main)> a || true
=> true
[3] pry(main)> a = false
=> false
[4] pry(main)> a || true
=> true
```

To avoid a silly regression like that, how about we change this to use `#fetch` instead?

```ruby
metadata.params.fetch(:cache_resources, false)
```

File: lib/inspec/transport.rb
Comment (adamleff) [inline lib/inspec/transport.rb]: Unless it's necessary, I'd rather we didn't abbreviate variable names. (I know we just did it in the Train platforms PR - i.e. `plat` instead of `platform` - but it still bugs me 🙂 )

Any reason this shouldn't be `connection`?

File: lib/inspec/transport.rb
Comment (adamleff) [inline lib/inspec/transport.rb]: Can we provide an example or two in the comment as to the type of method calls we're catching here, deeming this necessary?

File: lib/inspec/transport.rb
Comment (adamleff) [inline lib/inspec/transport.rb]: These should be debug messages, no?

File: lib/inspec/transport.rb
Comment (adamleff) [inline lib/inspec/transport.rb]: Debug instead of warn?

File: test/unit/transport_test.rb
Comment (adamleff) [inline test/unit/transport_test.rb]: What's the purpose of assigning this to a new variable? This happens throughout the tests, so I'm not going to comment on all of them, but I'm curious as to why it's necessary.

File: test/unit/transport_test.rb
Comment (adamleff) [inline test/unit/transport_test.rb]: Should probably also check that `os[:name] == 'mock'` here to properly exercise the `mock_os` method.

File: test/unit/transport_test.rb
Comment (adamleff) [inline test/unit/transport_test.rb]: Isn't `cache_resources` an `attr_accessor`? I like to avoid setting/getting instance variables unless necessary. I think we can just change this to:

```ruby
transport.cache_resources = true
```

File: test/unit/transport_test.rb
Comment (adamleff) [inline test/unit/transport_test.rb]: Same as above re: using the attr_accessor on the transport.

**Review by adamleff — CHANGES_REQUESTED** [review]
This is a great first pass, @jquick! I have some questions and some cleanup items.

**Comment by jquick** [comment]
@adamleff - Thanks for the feedback. I think I have those items resolved.

**Comment by chris-rock** [comment]
This is a great improvement @jquick and a very good start for a discussion!

After thinking for a while, I came to the conclusion that it would be best to implement a CacheConnection in train that would take another connection as initialize argument. This would solve multiple problems:
- every train user can use it, not just inspec
- the decision is still up to the train user if he wants to use caching or not
- InSpec will just use the cache connection for normal execution, but not for shell

This also means we should clean up the simple caching as we have it in train: https://github.com/chef/train/blob/master/lib/train/transports/docker.rb#L73-L80

**Comment by jquick** [comment]
@chris-rock - Thanks for the feedback. I will move the caching to Train. How do you feel about keeping the toggle here to enable/disable on the profile level?

**Comment by adamleff** [comment]
Moving it to Train seems totally fine to me, too. I still think we need to keep the toggle here on the InSpec side to allow the InSpec profile author to opt-in to the caching, and the InSpec will use the proper Train transport accordingly.

File: lib/inspec/backend.rb
Comment (chris-rock) [inline lib/inspec/backend.rb]: I think you cannot remove the copyright, since it stays available. This was developed by Dom before it was acquired by Chef

File: lib/inspec/backend.rb
Comment (chris-rock) [inline lib/inspec/backend.rb]: I think we do not need this here anymore

File: lib/inspec/profile.rb
Comment (chris-rock) [inline lib/inspec/profile.rb]: I think we should not make the caching part of the inspec.yml since this is static content. From my perspective, this  needs to be a flag for inspec cli where by default caching is turned off

**Review by chris-rock — CHANGES_REQUESTED** [review]
Nice cleanup @jquick I think we should not leave it to the profile author if caching is enabled or not. This is a decision of the runtime. We may decided to leave caching of for now but for InSpec 2.0 we may turn this on by default.

**Comment by jquick** [comment]
This should be set. Flag was set to `--backend-cache`/`--no-backend-cache` as `--cache` was already taken for profile caching on exec. Open to any other name though.

File: lib/inspec/base_cli.rb
Comment (chris-rock) [inline lib/inspec/base_cli.rb]: i think we should just call it cache

Comment (chris-rock) [reply lib/inspec/base_cli.rb]: maybe we set the default here? `default: false,`

File: lib/inspec/backend.rb
Comment (adamleff) [inline lib/inspec/backend.rb]: Is there another place we can/should put the `enable_cache` calls? The way I read this, every time something in InSpec calls for the `backend` it will send these methods to the Train connection. That could happen a LOT.

Maybe we just need an enable/disable method on the backend itself that can be called once when appropriate?

Also, we're making an assumption here that caching is disabled by default in Train, which is half correct -- we always cache files, and don't provide a way to disable currently. Should this have an `else` statement where we disable the cache?

This probably warrants further discussion after standup... until Train is consistent with the default behavior of both files and commands, we may want to consider only manipulating the command cache for now and then deal with files later.

**Review by adamleff — CHANGES_REQUESTED** [review]
I'd like to discuss this PR in more depth, perhaps after standup today. Comments in-line. Also, we cannot merge this until Train has a released version that supports the enable/disable calls. At that time, this PR will need to be updated to include a gemspec update so we properly depend on a version of Train that supports the feature.

**Comment by jquick** [comment]
Please note: test will bomb until we upgrade train with the caching changes

File: lib/inspec/backend.rb
Comment (adamleff) [inline lib/inspec/backend.rb]: Could you explain why this can't be an `else`? It's not quite clear to me. And probably warrants a code comment explaining the condition (i.e. why the mock backend is special here).

**Review by adamleff — CHANGES_REQUESTED** [review]
One last question regarding the caching settings conditional and we should be good to go.

File: lib/inspec/backend.rb
Comment (jquick) [reply lib/inspec/backend.rb]: Yeah I'll add a comment. Basically the mock transport uses the base connection caching also. So we never want to disable it. We hard set the cache values for testing and usually bypass thor so it would always be false. It actually may be cleaner to put the mock check on the enable and just ensure its enabled.

Comment (adamleff) [reply lib/inspec/backend.rb]: What would happen in the mock transport if it was disabled? Slower tests? If so, by how much?

Comment (jquick) [reply lib/inspec/backend.rb]: Not really, you cant actually run file/command calls via mock without preloading it. Else it will just fall to errors. So for the core of MockTransport to work caching is always needed. Ex:
https://github.com/chef/train/blob/master/lib/train/transports/mock.rb#L139 

We hard code it during the transport creation but this would override it.

**Review by adamleff — APPROVED** [review]
Conditionally approved. Before we merge:

 * a new version of Train needs to be released
 * this PR needs to be updated with a gemspec pin update for the train gem
 * tests go green

Nice work, @jquick!

File: lib/inspec/backend.rb
Comment (chris-rock) [inline lib/inspec/backend.rb:65]: should we explicitly enable file caching? so we do not depend on train defaults?

Comment (jquick) [reply lib/inspec/backend.rb:65]: added

**Review by chris-rock — APPROVED** [review]
Great improvement @jquick

**Comment by jquick** [comment]
This is tested and looks good. We wont have as good windows caching until https://github.com/chef/inspec/pull/2368 is merged. It would be nice to rebase this again after that is merged and make sure my numbers match up.

**Comment by jquick** [comment]
Actually hold on this, working on the helper fix for the train, its causing the detect to run for mocks.

**Comment by jquick** [comment]
Confirmed the cache timings on various windows and unix systems. This looks good.

**Pull request merged**
